### PR TITLE
Reverse lookup fixes and enhancements.

### PR DESCRIPTION
**Pull request by gianm**
Enhancements to the reverse-lookup optimization that is part of InDimFilter. As a follow-up to this, I'd like to add the optimization to the SQL planner. But it helps to improve the code before doing that. It has various problems that I think haven't been obvious because the optimization only applies to the `lookup` extractionFn (in core, this is only used for lookups inline with the query), not `registeredLookup` (used for lookups defined with the Coordinator API).

1) Add a "mayIncludeUnknown" parameter to DimFilter#optimize. This is important
   because otherwise the reverse-lookup optimization is done improperly when
   the "in" filter appears under a "not", and the lookup extractionFn may return
   null for some possible values of the filtered column. The "includeUnknown" test
   cases in InDimFilterTest illustrate the difference in behavior.

2) Enhance InDimFilter#optimizeLookup to handle "mayIncludeUnknown", and to be able
   to do a reverse lookup in a wider variety of cases.

3) Make "unapply" protected in LookupExtractor, and move callers to "unapplyAll".
   The main reason is that MapLookupExtractor, a common implementation, lacks a
   reverse mapping and therefore does a scan of the map for each call to "unapply".
   For performance sake these calls need to be batched.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java [hunk 1]
```diff
@@ -227,22 +228,32 @@ public byte[] getCacheKey()
...
-          inFilter.dimension,
-          inFilter.values.iterator().next(),
-          inFilter.getExtractionFn(),
+          dimension,
+          newValues.iterator().next(),
+          newExtractionFn,
           filterTuning
       );
+    } else if (newValues == values) {
+      return this;
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java:253]: should we just bail early with this if extractionFn is null?

File: processing/src/main/java/org/apache/druid/query/lookup/LookupExtractor.java [hunk 2]
```diff
@@ -68,37 +72,30 @@ public Map<String, String> applyAll(Iterable<String> keys)
...
+   * @param values set of values to reverse lookup. {@link NullHandling#nullToEmptyIfNeeded(String)} will have
+   *               been applied to each value.
+   *
+   * @return iterator of keys that map to to the provided set of values. May contain duplicate keys. Returns null if
+   * this lookup instance does not support reverse lookups.
    */
-
-  public Map<String, List<String>> unapplyAll(Iterable<String> values)
+  @Nullable
+  public Iterator<String> unapplyAll(Set<String> values)
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/lookup/LookupExtractor.java:96]: At first I was thinking if we are changing this interface, maybe we should also consider changing `unapply` itself to also spit out an iterator. However, looking closer since `MapLookupExtractor` overrides this it probably isn't a super big deal, and I'm not sure the other two implementations could actually offer much of an improvement if unapply was lazier, I was mostly just thinking that it is kind of sad this default implementation spits out an iterator but has to materialize a list for each `unapply` so it can get an iterator. 

tl;dr i think this is fine and nothing needs to change

**Review by clintropolis — APPROVED** [review]

File: processing/src/main/java/org/apache/druid/query/lookup/LookupExtractor.java (hunk 2 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/query/lookup/LookupExtractor.java:96]: I had a similar thought when writing it. It turns out that none of the implementations would benefit from laziness.

File: processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java (hunk 1 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java:253]: I think no, because the checks for `values` size are still needed to turn the filter into a `false` or `selector` filter if it has zero or one elements. (Even with no `extractionFn`)
